libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Encoding/ToCBOR.hs

```diff
@@ -0,0 +1,709 @@
+import Codec.CBOR.ByteArray (ByteArray (..))
+import Codec.CBOR.ByteArray.Sliced (SlicedByteArray (SBA), fromByteArray)
+import Codec.CBOR.Term (Term (..))
+import Control.Category (Category ((.)))
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BS.Lazy
+import qualified Data.ByteString.Short as SBS
+import qualified Data.ByteString.Short.Internal as SBS
+import Data.Fixed (Fixed (..), Nano, Pico)
+import Data.Foldable (toList)
+import Data.Functor.Foldable (cata, project)
+import Data.IP (IPv4, IPv6)
+import Data.Int (Int16, Int32, Int64, Int8)
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.Map as Map
+import qualified Data.Maybe.Strict as SMaybe
+import qualified Data.Primitive.ByteArray as Prim (ByteArray (..))
+import Data.Ratio (Ratio)
+import qualified Data.Sequence as Seq
+import qualified Data.Sequence.Strict as SSeq
+import qualified Data.Set as Set
+import Data.Tagged (Tagged (..))
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Text.Lazy.Builder (Builder)
+import Data.Time.Clock (NominalDiffTime, UTCTime (..))
+import Data.Typeable (Proxy (..), TypeRep, Typeable, typeRep)
+import qualified Data.VMap as VMap
+import qualified Data.Vector as V
+import qualified Data.Vector.Primitive as VP
+import qualified Data.Vector.Storable as VS
+import qualified Data.Vector.Unboxed as VU
+import Data.Void (Void, absurd)
+import Data.Word (Word16, Word32, Word64, Word8)
+import Foreign.Storable (sizeOf)
+import Formatting (bprint, build, shown, stext)
+import qualified Formatting.Buildable as B (Buildable (..))
+import Numeric.Natural (Natural)
+import Prelude hiding (encodeFloat, (.))
+#if MIN_VERSION_recursion_schemes(5,2,0)
+import Data.Fix (Fix(..))
+#else
+import Data.Functor.Foldable (Fix(..))
+#endif
+
+class Typeable a => ToCBOR a where
+  toCBOR :: a -> Encoding
+
+  encodedSizeExpr :: (forall t. ToCBOR t => Proxy t -> Size) -> Proxy a -> Size
+  encodedSizeExpr = todo
+
+  encodedListSizeExpr :: (forall t. ToCBOR t => Proxy t -> Size) -> Proxy [a] -> Size
+  encodedListSizeExpr = defaultEncodedListSizeExpr
+
+-- | A type used to represent the length of a value in 'Size' computations.
+newtype LengthOf xs = LengthOf xs
+
+instance Typeable xs => ToCBOR (LengthOf xs) where
+  toCBOR = error "The `LengthOf` type cannot be encoded!"
+
+-- | Default size expression for a list type.
+defaultEncodedListSizeExpr ::
+  forall a.
+  ToCBOR a =>
+  (forall t. ToCBOR t => Proxy t -> Size) ->
+  Proxy [a] ->
+  Size
+defaultEncodedListSizeExpr size _ =
+  2 + size (Proxy @(LengthOf [a])) * size (Proxy @a)
+
+newtype PreEncoded = PreEncoded {unPreEncoded :: BS.ByteString}
+
+instance ToCBOR PreEncoded where
+  toCBOR = encodePreEncoded . unPreEncoded
+
+--------------------------------------------------------------------------------
+-- Size expressions
+--------------------------------------------------------------------------------
+
```

Comment:
We should ask around if anyone uses the size stuff, and get rid of it if not.  Maybe we should just deprecate it at first.

Reply:
This is a good idea. I'd rather create a separate task for this, since I do have a suspicion that it might be used in some obscure cases. 

Here is the ticket: https://github.com/input-output-hk/cardano-ledger/issues/3071